app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/CoroutineObserver.kt

```diff
@@ -0,0 +1,48 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2020 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package org.kiwix.kiwixmobile.localFileTransfer
+
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.OnLifecycleEvent
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
+
+class CoroutineObserver @Inject constructor() : LifecycleObserver, CoroutineScope {
+  private lateinit var job: Job
+  override val coroutineContext: CoroutineContext
+    get() = job + Dispatchers.Main
+
+  init {
+    onCreate()
+  }
+
+  @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
+  fun onCreate() {
+    job = Job()
+  }
```

Comment:
Here I thought it automatically creates job but it didn't. I had to initiate in the constructor. Why is that?

Reply:
okay... maybe I added to the confusion with my explanation. You are creating a new scope and tying this to the lifecycle of the fragment but fragment already has a scope

![image](https://user-images.githubusercontent.com/3358282/98224670-a7040980-1f4b-11eb-9e96-9420edac8c63.png)

so this is overkill. Much like how we pass in listeners to the wifidirectManager I am happy to just pass this in.

Additionally there are 2 instances of this class, 1 in the fragment  and 1 in the WifiDirectManager so a cancel will do nothing.

On the original question: I think it does, registering in onCreate should mean that `@OnLifecycleEvent(Lifecycle.Event.ON_CREATE) onCreate` gets called however this comes back to the "there are 2 instances" problem above

Reply:
@macgills is [this](https://github.com/kiwix/kiwix-android/pull/2389/commits/07a9d426c04c59fab51d3315e6b9935b008a1216) the better way to cancel coroutines?

Reply:
no, the best way is to let someone else do it.
let me rewrite my screenshot
```
lifecycleScope.launch{
  delay(ONE_MILLION_YEARS)
 println("hi gouri")
}
```

that is all I need to do to ensure this gets cancelled when this fragment gets destroyed sometime in the next million years. Pass in this scope to wifiDirectManager and use it to launch coroutines

Reply:
done